dataplane/linux/wireguard_mgr.go

```diff
@@ -0,0 +1,104 @@
+// limitations under the License.
+
+package intdataplane
+
+import (
+	log "github.com/sirupsen/logrus"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+
+	"github.com/projectcalico/felix/ip"
+	"github.com/projectcalico/felix/wireguard"
+
+	"github.com/projectcalico/felix/proto"
+)
+
+// wireguardManager manages the dataplane resources that are used for wireguard encrypted traffic. This includes:
+// -  Routing rule to route to the wireguard routing table
+// -  Route table and rules specifically to handle routing to the wireguard interface, or to return to default routing
+//    (depending on whether the remote node supports wireguard)
+// -  Wireguard interface lifecycle
+// -  Wireguard peer configuration
+//
+// The wireguard component implements the routetable interface and so dataplane programming is triggered through calls
+// to the Apply method, with period resyncs occuring after calls to QueueResync. Calls from the main OnUpdate method
+// call through to the various update methods on the wireguard module which simply record state without actually
+// programming.
+type wireguardManager struct {
+	// Our dependencies.
+	wireguardRouteTable *wireguard.Wireguard
+}
+
+type WireguardStatusUpdateCallback func(ipVersion uint8, id interface{}, status string)
+
+func newWireguardManager(
+	wireguardRouteTable *wireguard.Wireguard,
+) *wireguardManager {
+	return &wireguardManager{
+		wireguardRouteTable: wireguardRouteTable,
+	}
+}
+
+func (m *wireguardManager) OnUpdate(protoBufMsg interface{}) {
+	log.WithField("msg", protoBufMsg).Debug("Received message")
+	switch msg := protoBufMsg.(type) {
+	case *proto.HostMetadataUpdate:
+		log.WithField("msg", msg).Debug("HostMetadataUpdate update")
+		m.wireguardRouteTable.EndpointUpdate(msg.Hostname, ip.FromString(msg.Ipv4Addr))
+	case *proto.HostMetadataRemove:
+		log.WithField("msg", msg).Debug("HostMetadataRemove update")
+		m.wireguardRouteTable.EndpointRemove(msg.Hostname)
+	case *proto.RouteUpdate:
+		log.WithField("msg", msg).Debug("RouteUpdate update")
+		if msg.Type != proto.RouteType_REMOTE_WORKLOAD {
+			log.Debug("RouteUpdate is not a peer workload update, ignoring")
+			return
+		}
+		cidr := ip.MustParseCIDROrIP(msg.Dst)
+		if cidr != nil {
+			m.wireguardRouteTable.EndpointAllowedCIDRAdd(msg.DstNodeName, cidr)
+		}
+	case *proto.RouteRemove:
+		log.WithField("msg", msg).Debug("RouteRemove update")
+		cidr := ip.MustParseCIDROrIP(msg.Dst)
+		if cidr != nil {
+			m.wireguardRouteTable.EndpointAllowedCIDRRemove(cidr)
+		} else {
+			log.Error("error parsing RouteRemove CIDR", msg.Dst)
+		}
+	case *proto.WireguardEndpointUpdate:
+		log.WithField("msg", msg).Debug("WireguardEndpointUpdate update")
+		key, err := wgtypes.ParseKey(msg.PublicKey)
+		if err != nil {
+			log.WithError(err).Errorf("error parsing wireguard public key %s for node %s", msg.PublicKey, msg.Hostname)
+		}
+		ifaceAddr := ip.FromString(msg.InterfaceAddr)
```

Comment:
What does a nil address mean here?  Should the calc graph convert nil updates to deletions or is the other data in the update still useful?

Reply:
it just means an ip address will not be assigned to the wg interface itself.  AFAICT this is fine for pod to pod traffic - I think the wg interface is more useful to handle host to host encryption where you can't simply send packets to the host IP.